_posts/2021-09-03-rudiments-of-data-wrangling-in-rust.md

```diff
@@ -0,0 +1,514 @@
+common situations. For example, initializing a vector with
+the `vec!` macro lets you focus on the initial contents instead
+of the code that allocates memory and adds each individual 
+element.
+
+The `println!` macro also takes care of formatting the data.
+You can let Rust print a value like it pleases, but you can
+also utilize the comprehensive formatting templates (also used
+by the `format!` macro).
+
+Since I stored the range of the interesting years earlier,
+I can use it when I present the results. First I print out a
+heading:
+
+    println!("EV registrations {}-{} by year:", 
+        year_range.start(), year_range.end());
+
+Then, because I'm confident that I have an entry in the hashmap
+for each of the years (remember, if there wasn't already an entry,
+I made one, and initialized it to zero), I can just loop through
+the range and print out the result for each year.
+
+    for year in year_range {
+        println!("{:>5}: {:>5}", 
+            year, registrations.get(&year).unwrap());
+    }
+
+Getting an entry from the hashmap can fail, but in this case 
+I just use the common Rust technique of "unwrapping" the result.
+Keep in mind that if you're not writing a quick and dirty utility
+like this one, unwrapping is not the best thing to do, because
+failure will cause the program to panic. Using unwrap is somewhat
+opinion based, but in a case like this I think it can be justified.
+
+Finally, I can print the total number.
+
+    println!("Total: {:>5}", electric_car_count);
+
+Just by eyeballing the numbers you can see that the growth 
+in EV registrations has been quite rapid. I could have also 
+computed the count while printing out the results, instead
+of collecting it as we progressed through the records.
+
+## There's plenty more where that came from
+
+This has been a whirlwind tour of some common real-world tasks
+involved in writing a Rust program. However, even in this small
+program there are many Rust features that I was not able to
+explain in depth, such as borrowing. (Even lifetimes already made a 
+brief appearance!)
+
+Also, you could easily add a visualization of the registration
+counts using the [`plotters`](https://crates.io/crates/plotters) crate, and extend the analysis of the
+vehicle data using the [`ndarray`](https://crates.io/crates/ndarray) crate.
+
+As for performance, it would be interesting to compare
+of this program and an equivalent Python solution in terms
+of speed, but also in terms of ergonomics. When you're first
+starting out with any programming language it takes a lot
+of time to look things up and assemble the parts to a 
+working whole. With Rust the learning curve can be steep,
+but I think it's time well invested.
+
+Here are the three biggest takeways for anyone considering
+a deeper dive into Rust:
+
+* The borrow checked is your friend.
+* Embrace the crates.
+* Algebraic data types FTW!
+
+Hope you enjoy wrangling data, or anything else, in Rust!
```

Comment:
Good structuring to small segments, perfect ending.

Always good to see if you can link other content to keep people on site. I see we have one previous Rust related article:

https://dev.solita.fi/2019/11/21/enforcing-database-transactions-with-rust.html